### ISSUE TITLE
Version 3.0 breaks builds on certain applications

### ISSUE THREAD

**Issue by kekePower** [sphinx-doc/sphinx#7438]
**Describe the bug**
I built [Botan](https://github.com/randombit/botan/) on Mageia Cauldron x86_64 and the build failed when using Sphix to build the documentation. Using Doxygen worked fine.

You can [see the full build log](https://github.com/randombit/botan/issues/2324) and my bug report over at Botan.

**To Reproduce**
Steps to reproduce the behavior: Build Botan using `--with-sphinx` and see that it fails.
```
$ git clone https://github.com/randombit/botan
$ cd botan
$ ./configure.py --with-sphinx
$ make
```

**Expected behavior**
I expect Sphinx to build the docs successfully.

**Environment info**
- OS: Mageia Cauldron x86_64
- Python version: 3.8.2
- Sphinx version: 3.0.0
- Sphinx extensions:  N/A
- Extra tools: N/A

**Additional context**
Full build log: https://raw.githubusercontent.com/kekePower/mmbl/6ddfe29086def989b9867c1893548acd716a5939/2020/04/07/20%3A07%3A51/botan2-2.14.0-1.mga8.src.rpm/build.0.20200407175517.log

**Comment by jakobandersen** [comment]
Thanks for reporting! It looks like a tricky bug with the merging of data that only happens when using parallel building (``-j 2`` or more), and with a specific order of merging.

**Comment by jakobandersen** [comment]
This should be fixed in branch 3.0.x (future v3.0.1).
